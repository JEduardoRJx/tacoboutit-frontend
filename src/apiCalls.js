export const getRestaurants = async (lat = 39.7392, lng = -104.9903) => {
  const response = await fetch(`https://tacoboutit-backend.herokuapp.com/api/v1/restaurants/retrieve/?lat=${lat}&lng=${lng}`);
  if(!response.ok) {
    throw Error('Failed to fetch restaurants near you');
  }
  return response.json();
}

export const newTaco = async (type, restaurant) => {
  const response = await fetch('https://tacoboutit-backend.herokuapp.com/api/v1/tacos/new/', {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({ type, restaurant }),
  });
  if(!response.ok) {
    return Error('Failed to add taco.');
  }
  const resp = await response.json();
  return resp;
}

export const addReview = async (tacoId, rating, review = 'No Review Available') => {
  const response = await fetch(`https://tacoboutit-backend.herokuapp.com/api/v1/reviews/`, {
    method: 'POST', 
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({
      taco: tacoId,
      rating,
      review,
    }),
  })
  if(!response.ok) {
    return Error('Failed to add review');
  }
  const resp = await response.json();
  return resp;
}
