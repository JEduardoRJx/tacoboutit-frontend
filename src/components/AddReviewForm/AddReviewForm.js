import React, { Component } from 'react'
import { Text, View, Image, StyleSheet, Dimensions, Linking, TouchableOpacity, ScrollView, Modal, Button, Picker, TextInput } from 'react-native';
import { addReview } from '../../apiCalls'

export default class AddReviewForm extends Component {
  state = {
    tacoId: this.props.tacos[0].id,
    rating: 1,
    review: '',
    type: this.props.tacos[0].type, 
    error: ''
  }

  renderTacoTypes = () => {
    const { tacos } = this.props
    return tacos.map(taco => <Picker.Item label={taco.type} value={taco.type.toLowerCase()} key={taco.id} />)
  }

  renderRatings = () => {
    let ratings = [];
    for(let i = 1; i <= 10; i++) {
      if (i !== 10) {
        ratings.push(<Picker.Item label={String(i)} value={i} key={i}/>)
        ratings.push(<Picker.Item label={String(i + .5)} value={i + .5} key={i}/>)
        
      } else {
        ratings.push(<Picker.Item label={String(i)} value={i} key={i}/>)
      }
    }
    return ratings
  }

  handleTacoTypeChange = (itemValue, itemIndex) => {
    const id = this.props.tacos[itemIndex].id
    this.setState({tacoId: id, type: itemValue})
  }

  handleRatingChange = (itemValue) => {
    this.setState({rating: Number(itemValue)})
  }

  submitNewReview = async () => {
    let { tacoId, rating, review } = this.state
    let response;
    const { updateLocalReviews } = this.props
    if(this.state.review === '') {
      response = await addReview(tacoId, rating)
      if(!response.ok) {
        this.setState({error: 'Failed to post review'})
      }
    } else {
      response = await addReview(tacoId, rating, review)
      if(!response.ok) {
        this.setState({error: 'Failed to post review'})
      }
    }
    updateLocalReviews(response)
  }

  render() {
    return (
      <View style={{flex: 2}}>
        <Text>Add Review</Text>
        <Text>Type:</Text>
        <Picker
          onValueChange={(itemValue, itemIndex) => this.handleTacoTypeChange(itemValue, itemIndex)}>{this.renderTacoTypes()}</Picker>
        <Text>Rating:</Text>
        <Picker 
        selectedValue={this.state.rating}
        onValueChange={(itemValue) => this.handleRatingChange(itemValue)}
        >{this.renderRatings()}</Picker>
        <Text>Review:</Text>
        <TextInput
          multiline={true}
          numberOfLines={4}
          maxLength={50}
          placeholder='Add A Review'
          onChangeText={(text) => {
            this.setState({review: text})
          }}
          />
        <Button onPress={() => this.submitNewReview()} title='Submit' />
      </View>
    )
  }
}