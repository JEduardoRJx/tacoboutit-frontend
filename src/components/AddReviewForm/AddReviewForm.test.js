import React from 'react';
import AddReviewForm from './AddReviewForm';
import { shallow } from 'enzyme';
import * as api from '../../apiCalls';

describe('AddReviewForm', () => {
  const mockToggleReviewForm = jest.fn();
  const mockTacos = [
    {
        "id": 9,
        "type": "Testing testing",
        "restaurant_id": 2,
        "average_rating": 5.0,
        "reviews": [
            {
                "id": 1,
                "rating": 7.0,
                "review": "its ok",
                "taco": 9
            },
            {
                "id": 2,
                "rating": 4.0,
                "review": "lame",
                "taco": 9
            },
            {
                "id": 3,
                "rating": 4.0,
                "review": null,
                "taco": 9
            },
            {
                "id": 4,
                "rating": 1.0,
                "review": "This taco put my son in the hospital",
                "taco": 9
            },
            {
                "id": 5,
                "rating": 10.0,
                "review": "The best",
                "taco": 9
            },
            {
                "id": 6,
                "rating": 4.0,
                "review": "meh",
                "taco": 9
            }
        ]
    }
];
  let wrapper;
  const mockUpdateLocalReviews = jest.fn();
  beforeEach(() => {
    wrapper = shallow(<AddReviewForm 
      tacos={mockTacos} 
      updateLocalReviews={mockUpdateLocalReviews}
      toggleReviewForm={mockToggleReviewForm}
    />);
  })
  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('handleRatingChange should update state with a new rating', () => {
    expect(wrapper.state('rating')).toEqual(1);
    wrapper.instance().handleRatingChange(5.5);
    expect(wrapper.state('rating')).toEqual(5.5);
  });

  it('handleTacoTypeChange should update props tacoId and type in state', () => {
    const initialExpectedState = {"rating": 1, "review": "", "tacoId": 9, "type": "Testing testing", error: "" };
    expect(wrapper.state()).toEqual(initialExpectedState);
    const expectedState = {"rating": 1, "review": "", "tacoId": 9, "type": "new type", error: "" };
    wrapper.instance().handleTacoTypeChange('new type', 0);
    expect(wrapper.state()).toEqual(expectedState);
  });
  it('submitTacoReview should post new review to the backend and update local reviews when review is missing', async () => {
    api.addReview = jest.fn().mockImplementation(() => {
      return Promise.resolve({ ok: true });
    });
    await wrapper.instance().submitNewReview();
    expect(api.addReview).toHaveBeenCalledWith(9, 1);
    expect(mockUpdateLocalReviews).toHaveBeenCalledWith({ ok: true });
  });
  it('submitTacoReview should post new review to the backend and update local reviews when review is present', async () => {
    api.addReview = jest.fn().mockImplementation(() => {
      return Promise.resolve({ ok: true });
    });
    wrapper.setState({ review: 'This is a test review' });
    await wrapper.instance().submitNewReview();
    expect(api.addReview).toHaveBeenCalledWith(9, 1, 'This is a test review');
    expect(mockUpdateLocalReviews).toHaveBeenCalledWith({ ok: true });
  });
  it('submitTacoReview should set error in state if adding a new review fails', async () => {
    api.addReview = jest.fn().mockImplementation(() => Promise.resolve({ ok: false }));
    await wrapper.instance().submitNewReview();
    expect(wrapper.state('error')).toEqual('Failed to post review');
    wrapper.setState({ review: 'This is a test review', error: '' });
    expect(wrapper.state('error')).toEqual('');
    await wrapper.instance().submitNewReview();
    expect(wrapper.state('error')).toEqual('Failed to post review');
  });
});