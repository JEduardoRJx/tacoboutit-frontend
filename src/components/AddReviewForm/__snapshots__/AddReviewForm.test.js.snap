// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddReviewForm should match the snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#DCDCDC",
      "flex": 4,
      "justifyContent": "space-around",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1.25,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 16,
            "textAlign": "center",
          }
        }
      >
        Select Taco Type
      </Text>
      <Picker
        mode="dialog"
        onValueChange={[Function]}
        selectedValue="Testing testing"
      >
        <PickerItem
          key="9"
          label="Testing testing"
          value="testing testing"
        />
      </Picker>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 16,
            "textAlign": "center",
          }
        }
      >
        Select Rating
      </Text>
      <Picker
        mode="dialog"
        onValueChange={[Function]}
        selectedValue={1}
      >
        <PickerItem
          key="1"
          label="1"
          value={1}
        />
        <PickerItem
          key="1"
          label="1.5"
          value={1.5}
        />
        <PickerItem
          key="2"
          label="2"
          value={2}
        />
        <PickerItem
          key="2"
          label="2.5"
          value={2.5}
        />
        <PickerItem
          key="3"
          label="3"
          value={3}
        />
        <PickerItem
          key="3"
          label="3.5"
          value={3.5}
        />
        <PickerItem
          key="4"
          label="4"
          value={4}
        />
        <PickerItem
          key="4"
          label="4.5"
          value={4.5}
        />
        <PickerItem
          key="5"
          label="5"
          value={5}
        />
        <PickerItem
          key="5"
          label="5.5"
          value={5.5}
        />
        <PickerItem
          key="6"
          label="6"
          value={6}
        />
        <PickerItem
          key="6"
          label="6.5"
          value={6.5}
        />
        <PickerItem
          key="7"
          label="7"
          value={7}
        />
        <PickerItem
          key="7"
          label="7.5"
          value={7.5}
        />
        <PickerItem
          key="8"
          label="8"
          value={8}
        />
        <PickerItem
          key="8"
          label="8.5"
          value={8.5}
        />
        <PickerItem
          key="9"
          label="9"
          value={9}
        />
        <PickerItem
          key="9"
          label="9.5"
          value={9.5}
        />
        <PickerItem
          key="10"
          label="10"
          value={10}
        />
      </Picker>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "flex": 1,
        "paddingHorizontal": 20,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      maxLength={50}
      multiline={true}
      numberOfLines={4}
      onChangeText={[Function]}
      placeholder="Add A Review"
      rejectResponderTermination={true}
      style={
        Object {
          "fontSize": 16,
        }
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#DCDCDC",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-around",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
        style={
          Object {
            "alignContent": "center",
            "backgroundColor": "#00BFFF",
            "borderRadius": 50,
            "height": "80%",
            "justifyContent": "center",
            "width": "90%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 20,
              "textAlign": "center",
            }
          }
        >
          Cancel
        </Text>
      </TouchableOpacity>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
        style={
          Object {
            "alignContent": "center",
            "backgroundColor": "#00BFFF",
            "borderRadius": 50,
            "height": "80%",
            "justifyContent": "center",
            "width": "90%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 20,
              "textAlign": "center",
            }
          }
        >
          Submit
        </Text>
      </TouchableOpacity>
    </View>
  </View>
</View>
`;
